feat: translations integration tests

bhearsum opened the pull request:
This adds support for running the translations training pipeline as part of integration tests. The first two commits, which add support for fetching upstream tasks that didn't run as part of a decision task and support for rewriting dependencies to the yet-to-be-scheduled tasks are the interesting parts. The other few commits are pretty boring, and simply fix a minor bug and enable the translations tests.

Note that these depends on https://github.com/mozilla-releng/fxci-config/pull/228, a yet to be opened PR in https://github.com/mozilla/translations to add a pipeline cronjob. It's currently running against the staging repository and working well though.

bhearsum commented:
/taskcluster integration

bhearsum commented:
/taskcluster integration

bhearsum commented:
/taskcluster integration

bhearsum commented:
/taskcluster integration

bhearsum commented:
/taskcluster integration

bhearsum commented:
/taskcluster integration

bhearsum commented:
/taskcluster integration

bhearsum commented:
/taskcluster integration

bhearsum commented:
/taskcluster integration

bhearsum commented:
/taskcluster integration

bhearsum commented:
/taskcluster integration

bhearsum commented:
/taskcluster integration

bhearsum commented:
/taskcluster integration

bhearsum commented:
/taskcluster integration

bhearsum commented:
/taskcluster integration

bhearsum commented:
/taskcluster integration

bhearsum commented:
/taskcluster integration

bhearsum commented:
/taskcluster integration

bhearsum commented:
/taskcluster integration

bhearsum commented:
/taskcluster integration

bhearsum commented:
/taskcluster integration

bhearsum commented:
/taskcluster integration

bhearsum commented:
/taskcluster integration

bhearsum commented:
/taskcluster integration

bhearsum commented:
/taskcluster integration

ahal commented on taskcluster/fxci_config_taskgraph/transforms/integration_test.py:
```diff
@@ -198,6 +252,19 @@ def schedule_tasks_at_index(config, tasks):
     for task in tasks:
+        include_attrs = task.pop("include-attrs", {})
+        exclude_attrs = task.pop("exclude-attrs", {})
+        include_deps = task.pop("include-deps", [])
```
This is getting complex enough that's it's probably worth adding a schema.

bhearsum replied:
Good call....

bhearsum commented:
/taskcluster integration

bhearsum commented:
/taskcluster integration

bhearsum commented:
@ahal - This is finally in a reviewable state. I've updated the description further up, and the commits should be pretty organized and straightforward to look at. https://stage.taskcluster.nonprod.cloudops.mozgcp.net/tasks/groups/clZrmWzyQZy-EZY-rRZ1jw is the most recent integration tests that I've run (nothing other than rebasing has changed since I ran those).

bhearsum commented:
Augh, I busted the latest training run by force pushing here...it almost run to completion though, and I have plenty of other runs that did. I can rerun it again if that's useful...

ahal commented on taskcluster/kinds/integration-test/kind.yml:
```diff
@@ -23,3 +23,18 @@ tasks:
+  translations:
+    description: "Run translations pipeline integration tests"
+    decision-index-paths:
+      # TODO: switch to production translations repo
```
Did you mean to switch this prior to landing?

bhearsum replied:
Sorry, I forgot to comment on this! Yes, this will be switched. It's waiting on https://github.com/mozilla-releng/fxci-config/pull/228, https://github.com/mozilla/translations/pull/966, and a cron run of the pipeline once those land.

ahal commented on taskcluster/kinds/firefoxci-artifact/kind.yml:
```diff
@@ -41,3 +41,18 @@ tasks:
+  translations:
+    description: "Fetch artifacts from Firefox-CI for translations pipeline tests"
+    decision-index-paths:
+      # TODO: switch to production translations repo
```
Ditto

ahal commented on taskcluster/fxci_config_taskgraph/transforms/integration_test.py:
```diff
@@ -2,16 +2,68 @@
+            A list of index paths in the Firefox CI (production) Taskcluster
+            instance index whose created tasks should be rerun in the staging
+            instance (subject to the filtering provided to this transform).
+            """.lstrip(),
```
Nice, thanks for the descriptions!

ahal commented on taskcluster/fxci_config_taskgraph/transforms/integration_test.py:
```diff
@@ -2,16 +2,68 @@
+        ): [str],
+    }
+)
+
```
You forgot to do `transforms.add_validate(SCHEMA)` below.

ahal commented on taskcluster/fxci_config_taskgraph/transforms/integration_test.py:
```diff
@@ -119,6 +171,15 @@ def rewrite_docker_image(taskdesc: dict[str, Any]) -> None:
+    if isinstance(moz_fetches, str):
+        return json.loads(moz_fetches)
+    elif isinstance(moz_fetches, dict) and moz_fetches.get("task-reference"):
+        return json.loads(moz_fetches["task-reference"])
```
What circumstances can this happen? Not a problem, just curious.

bhearsum replied:
Good question actually....I definitely wrote this because I found one that was a dict at some point, but in retrospect it might have been me misconfiguring a test. I'll sanity check this; it would certainly be preferable to drop it.

bhearsum replied:
This is not needed, I removed it!

ahal commented on taskcluster/fxci_config_taskgraph/transforms/integration_test.py:
```diff
@@ -144,7 +205,51 @@ def rewrite_private_fetches(taskdesc: dict[str, Any]) -> None:
 
 
-def make_integration_test_description(task_def: dict[str, Any], name_prefix: str):
+def rewrite_mirrored_dependencies(
```
A docstring would be nice. From a glance, I don't know what a "mirrored" dependency is.

ahal reviewed (changes requested):
This is excellent! Nicely documented and tested.

Requesting changes for the missing schema validation and my questions about using the production translations repo.

bhearsum commented:
/taskcluster integration

bhearsum commented:
/taskcluster integration

bhearsum commented:
/taskcluster integration

bhearsum commented:
/taskcluster integration

bhearsum commented:
Encountered a couple of novel things while doing fixups; moving back to draft while I deal with them.

bhearsum commented:
/taskcluster integration

bhearsum commented:
/taskcluster integration

bhearsum commented:
/taskcluster integration

bhearsum commented:
/taskcluster integration

bhearsum commented:
@ahal - this is ready for review again, but I'm going to leave it in Draft until I'm able to update the new `kind` entries to point at production. I've fixed the things you noted, but I also want to call attention to the most recent two commits, which are workarounds for things that came up when retesting this.

I have integration tests running in https://stage.taskcluster.nonprod.cloudops.mozgcp.net/tasks/groups/Vb5IURGhS5mLQy5np8IOAg (the only change between what that's running from and the latest is some updates to comments and commit messages).

bhearsum commented:
/taskcluster integration

bhearsum commented:
/taskcluster integration

bhearsum commented:
> @ahal - this is ready for review again, but I'm going to leave it in Draft until I'm able to update the new `kind` entries to point at production. I've fixed the things you noted, but I also want to call attention to the most recent two commits, which are workarounds for things that came up when retesting this.
> 
> I have integration tests running in https://stage.taskcluster.nonprod.cloudops.mozgcp.net/tasks/groups/Vb5IURGhS5mLQy5np8IOAg (the only change between what that's running from and the latest is some updates to comments and commit messages).

I've updated to point at production, and a fresh round of integration tests are running in https://stage.taskcluster.nonprod.cloudops.mozgcp.net/tasks/groups/WQwObYW7RmOWREX_YUPpXQ. Nothing notable has changed since I wrote the quoted comment and now.

bhearsum commented:
/taskcluster integration

bhearsum commented:
/taskcluster integration

bhearsum commented:
/taskcluster integration

bhearsum commented:
/taskcluster integration

bhearsum commented:
/taskcluster integration

bhearsum commented:
/taskcluster integration

bhearsum commented:
(Whoops, should've moved this back to draft awhile ago. Some tasks that expired uncovered some edge cases here that are proving tricky to work through.)

bhearsum commented:
/taskcluster integration

bhearsum commented:
/taskcluster integration

bhearsum commented:
/taskcluster integration

bhearsum commented:
/taskcluster integration

bhearsum commented:
/taskcluster integration

bhearsum commented:
/taskcluster integration

bhearsum commented:
Latest testing is having two issues:
- some of the training tasks appear to be getting _very_ aggressivelly spot killed; https://github.com/mozilla/translations/pull/1009 should help with this
- the windows tasks are not running at all - azure instances are getting requested but never spinning up

I'm going to try this all again on Monday.

bhearsum commented:
/taskcluster integration

bhearsum commented:
> Latest testing is having two issues:
> 
>     * some of the training tasks appear to be getting _very_ aggressivelly spot killed; [fix: use smaller instances for cron pipeline runs mozilla/translations#1009](https://github.com/mozilla/translations/pull/1009) should help with this

This turned out to be caused by https://github.com/mozilla-releng/fxci-config/pull/271.

>     * the windows tasks are not running at all - azure instances are getting requested but never spinning up

We're not sure what happened here; but there's no problem with them this time around.

ahal reviewed (approved):
Thanks, this looks reasonable to me!

I have some nits and one question I'm a bit confused on, but nothing worth a re-review.

bhearsum commented on taskcluster/fxci_config_taskgraph/transforms/integration_test.py:
```diff
@@ -128,7 +137,7 @@ def rewrite_private_fetches(taskdesc: dict[str, Any]) -> None:
 
     if "MOZ_FETCHES" in payload.get("env", {}):
-        fetches = json.loads(payload.get("env", {}).get("MOZ_FETCHES", "{}"))
+        fetches = load_fetches(payload.get("env", {}).get("MOZ_FETCHES", "{}"))
```
I don't think you're confused about anything here! This is just an oversight that wasn't caught by anything. I'll fix this up.

bhearsum commented:
/taskcluster integration
